Add guidelines for durations (#340)
Microsoft/api-guidelines: azure/Guidelines.md

:white_check_mark: **DO** use [RFC3339](https://datatracker.ietf.org/doc/html/rfc3339) for date/time.

> :white_check_mark: **DO** use a fixed time interval when using durations e.g., `ttlSeconds`. The unit of time should be part of the name, but must at least be documented in the property description.

Reviewer: Do we have some preference on the units? E.g. preferable second and millisecond?

In extreme case, what if service use [tick](https://docs.microsoft.com/en-us/dotnet/api/system.timespan.ticks?view=net-6.0)?

Reviewer 2: ```suggestion
:white_check_mark: **DO** use a fixed time interval when using durations e.g., `ttlSeconds`. The unit of time should be part of the name, e.g. `backupTimeInMinutes` 

✅ **DO** fully document the property and the interval in the description.
```

Any objection to being more direct on the guidance around documentation?
I might also add an example of the operation name.

Author: I could give a couple examples like I "e.g., `ttlSeconds` or `backupTimeInMinutes`. Should we just be explicit then that they should *always* include the unit?

@weidongxu-microsoft no, I don't feel we should be explicit about the time unit. It will vary from use to use, and if we made everything milliseconds, many models may need to use a 64-bit integer to at least parse the number where a 32-bit integer would do if in seconds, minutes, hours, etc.

Reviewer: Sure. Ask because I've just get caught by spring cloud service use int for cpu/memoryGB, then they decided to support 0.5 cpu and 0.5 GB :-(

> :white_check_mark: **DO** use a fixed time interval when using durations and include the time unit in the name e.g., `backupTimeInMinutes` or `ttlSeconds`. The time unit should also be included in the property description to support localization e.g., "How often to back up expressed in minutes," or "The lifetime of the resource expressed in seconds."

Reviewer 3: ```suggestion
:white_check_mark: **DO** use a fixed time interval to express durations and include the time unit in the property name e.g., `backupTimeInMinutes` or `ttlSeconds`. The time unit should also be included in the property description to support localization e.g., "How often to back up expressed in minutes," or "The lifetime of the resource expressed in seconds."
```

Reviewer 3: I don't know what you mean by "support localization"; please explain?

Author: The schema - models, properties, etc. - won't be localized, but the descriptions appear to be, at least when I look at https://docs.microsoft.com/de-de/rest/api/keyvault/keys/create-key/create-key. So I figure the unit of measure should also be in the description so it's localized, i.e. localizable.

Reviewer 3: I don't know how this is done (automated or manual) and if done manually, I don't know if this happens for all services. Regardless, we never mention anything about localization in any of our guidelines and so I'd prefer we not mention it here as mentioning it introduced questions and uncertainty.

> :heavy_check_mark: **YOU MAY** use [RFC3339](https://datatracker.ietf.org/doc/html/rfc3339) durations only when users must be able to specify a time interval that may change from month to month or year to year e.g., "P1Y" represents 366 days on a leap year.

Reviewer 4: Is that the right spec or did we mean ISO 8601?  The `Pwhatever` duration format is only discussed in the ABNF in the appendix and might not be as helpful as section 4.4.3 of ISO 8601?

Author: RFC 3339 supersedes ISO 8601, and the latter isn't accessible (paywalled) anywhere I can find outside of Wikipedia. So I don't think we can link to that, though I agree RFC 3339's ABNF isn't great. Should I link to Wikipedia as well / in lieu of?

Reviewer 4: Yeah, I'd suggest linking to Wikipedia as well.  I don't think most internal non-HTTP-experts would be successful with just RFC 3339.

Author: Okay, but I'm still going to reference RFC 3339, which we do in other places, like just right about this. I can link to wikipedia for more information. Maybe I just won't make RFC3339 a link.
